plugins/net_plugin/net_plugin.cpp

```diff
@@ -1868,16 +1867,18 @@ namespace eosio {
             // if starting to sync need to always start from lib as we might be on our own fork
             uint32_t lib_num = my_impl->get_chain_lib_num();
             sync_next_expected_num = std::max( lib_num + 1, sync_next_expected_num );
-            request_next_chunk( std::move(g) );
+            request_next_chunk( g.release() );
          }
       }
    }
 
-   connection_ptr sync_manager::find_next_sync_node() {
+   connection_ptr sync_manager::find_next_sync_node() REQUIRES(sync_mtx) {
       fc_dlog(logger, "Number connections ${s}, sync_next_expected_num: ${e}, sync_known_lib_num: ${l}",
               ("s", my_impl->connections.number_connections())("e", sync_next_expected_num)("l", sync_known_lib_num));
       deque<connection_ptr> conns;
-      my_impl->connections.for_each_block_connection([&](const auto& c) {
+      my_impl->connections.for_each_block_connection([sync_next_expected_num = sync_next_expected_num,
+                                                      sync_known_lib_num = sync_known_lib_num,
```

Comment:
Curious why you added the copy of these two members here.

Reply:
With the capture by reference:

```
      my_impl->connections.for_each_block_connection([&](const auto& c) {
         if (c->should_sync_from(sync_next_expected_num, sync_known_lib_num)) {
            conns.push_back(c);
         }
```

I get these errors:

```
/home/greg/github/enf/leap/plugins/net_plugin/net_plugin.cpp:1880:34: warning: reading variable 'sync_next_expected_num' requires holding mutex 'sync_mtx' [-Wthread-safety-analysis]
         if (c->should_sync_from(sync_next_expected_num, sync_known_lib_num)) {
                                 ^
/home/greg/github/enf/leap/plugins/net_plugin/net_plugin.cpp:1101:13: note: in instantiation of function template specialization 'eosio::sync_manager::find_next_sync_node()::(anonymous class)::operator()<std::shared_ptr<eosio::connection>>' requested here
            f(c);
            ^
/home/greg/github/enf/leap/plugins/net_plugin/net_plugin.cpp:1879:28: note: in instantiation of function template specialization 'eosio::connections_manager::for_each_block_connection<(lambda at /home/greg/github/enf/leap/plugins/net_plugin/net_plugin.cpp:1879:54)>' requested here
      my_impl->connections.for_each_block_connection([&](const auto& c) {
                           ^
/home/greg/github/enf/leap/plugins/net_plugin/net_plugin.cpp:1880:58: warning: reading variable 'sync_known_lib_num' requires holding mutex 'sync_mtx' [-Wthread-safety-analysis]
         if (c->should_sync_from(sync_next_expected_num, sync_known_lib_num)) {
                                                         ^
```

and unfotunately adding `REQUIRES(sync_mtx) ` does not fix it.